wkcuber/api/TiffData/TiffMag.py

```diff
@@ -48,96 +142,120 @@ def detect_value(
-            for file_name in self.list_files()
-        ]
+        available_tiffs = list(itertools.product(x_range, y_range, z_range))
 
-        for z in z_range:
-            self.tiffs[z] = TiffReader.open(
-                self.get_file_name_for_layer(z)
+        for xyz in available_tiffs:
+            self.tiffs[xyz] = TiffReader.open(
+                self.get_file_name_for_layer(xyz)
             )  # open is lazy
 
     def read(self, off, shape) -> np.array:
-        if not self.has_only_one_channel():
-            # modify the shape to also include the num_channels
-            shape = tuple(shape) + tuple([self.num_channels])
+        # if not self.has_only_one_channel():
+        # modify the shape to also include the num_channels
+        shape = tuple(shape) + tuple([self.header.num_channels])
+
+        data = np.zeros(shape=shape, dtype=self.header.dtype)
+        for (
+            xyz,
+            _,
+            offset_in_output_data,
+            offset_in_input_data,
+        ) in self.calculate_relevant_slices(off, shape):
+            x, y, z = xyz
+            z_index_in_data = z - off[2]
+
+            if xyz in self.tiffs:
+                # load data and discard the padded data
+                loaded_data = np.array(self.tiffs[xyz].read(), self.header.dtype)[
+                    offset_in_output_data[0] :, offset_in_output_data[1] :
+                ]
 
-        data = np.zeros(shape=shape, dtype=self.dtype)
-        for i, (z, offset, size) in enumerate(
-            self.calculate_relevant_slices(off, shape)
-        ):
-            if z in self.tiffs:
-                data[:, :, i] = np.array(self.tiffs[z].read(), self.dtype)[
-                    offset[0] : offset[0] + size[0], offset[1] : offset[1] + size[1]
+                index_slice = [
+                    slice(
+                        offset_in_input_data[0],
+                        offset_in_input_data[0] + loaded_data.shape[0],
+                    ),
+                    slice(
+                        offset_in_input_data[1],
+                        offset_in_input_data[1] + loaded_data.shape[1],
+                    ),
+                    z_index_in_data,
                 ]
-            else:
-                shape_without_z = shape[:2] + shape[3:]
-                data[:, :, i] = np.zeros(shape_without_z, self.dtype)
+                if self.has_only_one_channel():
+                    index_slice.append(0)
 
-        if self.has_only_one_channel():
-            # convert data into shape with dedicated num_channels (len(data.shape) == 4)
-            # this only effects data where the num_channel is 1 and therefore len(data.shape) was 3
-            # this makes it easier to handle both, multi-channel and single-channel, similar
-            data = np.expand_dims(data, 3)
+                # store the loaded data at the right position in 'data'
+                data[tuple(index_slice)] = loaded_data
```

Comment:
The array `data` has always the shape `(x, y, z, num_channel)` (even if num_channel is 1). However, the shapes of `loaded_data` differ depending on the number of channels. For single-channel data the shape of `loaded data` looks like `(x, y)`, if the data has multiple channels, the shape looks like `(x, y, num_channels)` (note that in both cases there is no z-dimension because an image belongs only to one z-dimension).

The code
```
if self.has_only_one_channel():
   index_slice.append(0)
```
is necessary to handle this mismatch of the shapes of `data` and `loaded_data` in case the single-channel is used.